File: kiko-api/src/services/judge/stageLayer.ts
/**
 * Stage Layer
 * Evaluates token lifecycle stage based on contract age
 * 
 * Stage definitions:
 * S0: Just created (< 10 minutes) - Extreme uncertainty
 * S1: Early (10 min - 2 hours) - High volatility
 * S2: Mid-term (2 - 24 hours) - Establishing pattern
 * S3: Stable (> 24 hours) - More predictable
 * S4: Post-Migration / Special stage
 */

import {
    StageLayer,
    TokenStage,
    LayerDecision,
    STAGE_THRESHOLDS,
} from '../../types/judgeTypes.js';

/**
 * Evaluate stage layer based on contract age
 * 
 * @param contractAgeMinutes - Age of contract in minutes
 * @param isMigrated - Whether token has been migrated (optional)
 * @returns StageLayer with stage, score, and decision
 */
export function evaluateStageLayer(
    contractAgeMinutes: number,
    isMigrated: boolean = false
): StageLayer {
    const reasons: string[] = [];

    let stage: TokenStage;
    let riskScore: number;
    let decision: LayerDecision;

    // Special case: Migrated token
    if (isMigrated) {
        stage = 'S4';
        riskScore = 0.5;
        decision = 'ALLOW_WITH_RISK';
        reasons.push('Token migrated - contract switch risk');
        reasons.push('Post-migration state needs extra verification');

        return {
            stage,
            contract_age_minutes: contractAgeMinutes,
            stage_risk_score: riskScore,
            stage_decision: decision,
            reasons,
        };
    }

    // Determine stage based on age
    if (contractAgeMinutes < 10) {
        // S0: Just created
        stage = 'S0';
        riskScore = 0.2;
        decision = 'ALLOW_WITH_RISK';
        reasons.push(`Just created (${contractAgeMinutes.toFixed(0)} minutes) - extreme uncertainty`);
        reasons.push('Price and liquidity may be highly volatile');
        reasons.push('Recommend a small test size only');

    } else if (contractAgeMinutes < 120) {
        // S1: Early (10 min - 2 hours)
        stage = 'S1';
        riskScore = 0.4;
        decision = 'ALLOW_WITH_RISK';
        const ageDisplay = contractAgeMinutes < 60
            ? `${contractAgeMinutes.toFixed(0)} minutes`
            : `${(contractAgeMinutes / 60).toFixed(1)} hours`;
        reasons.push(`Early stage (${ageDisplay}) - high volatility`);
        reasons.push('Smart money entries may create opportunities');

    } else if (contractAgeMinutes < 1440) {
        // S2: Mid-term (2 - 24 hours)
        stage = 'S2';
        riskScore = 0.6;
        decision = 'ALLOW';
        const hours = (contractAgeMinutes / 60).toFixed(1);
        reasons.push(`Mid stage (${hours} hours) - stabilizing`);
        reasons.push('Some trading history is available');

    } else {
        // S3: Stable (> 24 hours)
        stage = 'S3';
        riskScore = 0.8;
        decision = 'ALLOW';
        const days = (contractAgeMinutes / 1440).toFixed(1);
        reasons.push(`Stable stage (${days} days) - more predictable`);
        reasons.push('Lower risk but may miss early upside');
    }

    return {
        stage,
        contract_age_minutes: contractAgeMinutes,
        stage_risk_score: riskScore,
        stage_decision: decision,
        reasons,
    };
}

/**
 * Calculate contract age in minutes from creation timestamp
 */
export function calculateContractAgeMinutes(createdAt?: number | string): number {
    if (!createdAt) return 0;

    const createdTime = typeof createdAt === 'string'
        ? new Date(createdAt).getTime()
        : createdAt;

    const now = Date.now();
    const ageMs = now - createdTime;

    return Math.max(0, ageMs / (1000 * 60));
}

/**
 * Get stage label for display
 */
export function getStageLabel(stage: TokenStage): string {
    const labels: Record<TokenStage, string> = {
        'S0': 'Just created',
        'S1': 'Early',
        'S2': 'Mid',
        'S3': 'Stable',
        'S4': 'Special',
    };
    return labels[stage];
}

File: kikoapi/internal/services/judge/stage_layer.go
package judge

import (
	"fmt"

	"kikoapi/internal/types"
)

// EvaluateStageLayer evaluates token lifecycle stage based on contract age (minutes).
func EvaluateStageLayer(contractAgeMinutes int, isMigrated bool) types.StageLayer {
	var reasons []string
	var stage types.TokenStage
	var riskScore float64
	var decision types.LayerDecision

	if isMigrated {
		stage = types.StageS4
		riskScore = 0.5
		decision = types.LayerAllowWithRisk
		reasons = append(reasons, "Token migrated - contract switch risk", "Post-migration state needs extra verification")
		return types.StageLayer{
			Stage:            stage,
			ContractAgeMins:   contractAgeMinutes,
			StageRiskScore:   riskScore,
			StageDecision:    decision,
			Reasons:          reasons,
		}
	}

	if contractAgeMinutes < 10 {
		stage = types.StageS0
		riskScore = 0.2
		decision = types.LayerAllowWithRisk
		reasons = append(reasons, fmt.Sprintf("Just created (%d minutes) - extreme uncertainty", contractAgeMinutes), "Price and liquidity may be highly volatile", "Recommend a small test size only")
	} else if contractAgeMinutes < 120 {
		stage = types.StageS1
		riskScore = 0.4
		decision = types.LayerAllowWithRisk
		ageDisplay := fmt.Sprintf("%d minutes", contractAgeMinutes)
		if contractAgeMinutes >= 60 {
			ageDisplay = fmt.Sprintf("%.1f hours", float64(contractAgeMinutes)/60)
		}
		reasons = append(reasons, fmt.Sprintf("Early stage (%s) - high volatility", ageDisplay), "Smart money entries may create opportunities")
	} else if contractAgeMinutes < 1440 {
		stage = types.StageS2
		riskScore = 0.6
		decision = types.LayerAllow
		hours := float64(contractAgeMinutes) / 60
		reasons = append(reasons, fmt.Sprintf("Mid-term (%.1f hours) - establishing pattern", hours))
	} else {
		stage = types.StageS3
		riskScore = 0.8
		decision = types.LayerAllow
		reasons = append(reasons, fmt.Sprintf("Stable (%d+ hours) - more predictable", contractAgeMinutes/60))
	}

	return types.StageLayer{
		Stage:          stage,
		ContractAgeMins: contractAgeMinutes,
		StageRiskScore: riskScore,
		StageDecision:  decision,
		Reasons:        reasons,
	}
}

// CalculateContractAgeMinutes returns contract age in minutes from pairCreatedAt (unix ms).
// If nowUnixMs is 0, current time is not applied and 0 is returned when pairCreatedAt is set.
func CalculateContractAgeMinutes(pairCreatedAt *int64, nowUnixMs int64) int {
	if pairCreatedAt == nil || *pairCreatedAt <= 0 {
		return 0
	}
	if nowUnixMs <= 0 {
		return 0
	}
	diffMs := nowUnixMs - *pairCreatedAt
	if diffMs < 0 {
		return 0
	}
	return int(diffMs / (60 * 1000))
}
